app/thermal_management/thermalmgmt.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#include "thermalmgmt.h"
+#include "fan.h"
+#include "thermal_sensor.h"
+#include "board_config.h"
+#include "smc.h"
+#include "peci_hub.h"
+#include "pwrplane.h"
+
+LOG_MODULE_REGISTER(thermal, CONFIG_THERMAL_MGMT_LOG_LEVEL);
+
+/* PECI takes some time to initialize in CPU side, initial request will fail
+ * after reaching S0.
+ * Hence allow 1 sec time before reading temperature from peci.
+ * Till then report fail safe temperature of 28C.
+ */
+#define CPU_TEMP_ACCESS_DELAY_SEC		2U
+
+#define CPU_FAIL_SAFE_TEMPERATURE		28U
+
+/* CPU fail critical temperature value is 72C */
+#define CPU_FAIL_CRITICAL_TEMPERATURE		72U
+
+/*
+ * EC Self control fan speed based on CPU temperature info.
+ *
+ * Simpler approach used here is to control fan with liner profile.
+ * For CPU temperature between 0 to 100, fan to also rotate at equivalent speed,
+ * but rather than changing speed on every single degree change, it is changed
+ * after 8 degrees.
+ */
+#define GET_FAN_SPEED_FOR_TEMP(temp)		(temp & (~0x7))
+
+struct therm_sensor *therm_sensor_tbl;
+struct fan_dev *fan_dev_tbl;
+static u8_t max_adc_sensors;
+static u8_t max_fan_dev;
+static bool thermal_initialized;
+static bool peci_initialized;
+static u8_t fan_duty_cycle[FAN_DEV_TOTAL];
+static bool fan_duty_cycle_change;
+static int cpu_temp;
+/* Using module property, before introducing straps handler */
+static bool fan_override;
+
+static void init_fans(void)
+{
+	int ret;
+
+#ifdef CONFIG_THERMAL_FAN_OVERRIDE
+	fan_override = true;
+	LOG_WRN("Fan SW override enable: %d", fan_override);
+#endif
+
+	/* Get the list of fan devices supported for the board */
+	board_fan_dev_tbl_init(&max_fan_dev, &fan_dev_tbl);
+	LOG_DBG("Board has %d fans", max_fan_dev);
+
+	ret = fan_init(max_fan_dev, fan_dev_tbl);
+	if (ret) {
+		LOG_ERR("Failed to init fan");
+	}
+
+	fan_duty_cycle[FAN_CPU] = CONFIG_THERMAL_FAN_OVERRIDE_VALUE;
+	fan_duty_cycle_change = 1;
+}
+
+static void init_therm_sensors(void)
+{
+	uint8_t adc_ch_bits = 0;
+
+	board_therm_sensor_tbl_init(&max_adc_sensors, &therm_sensor_tbl);
+
+	LOG_INF("Num of thermal sensors: %d", max_adc_sensors);
+
+	for (uint8_t idx = 0; idx < max_adc_sensors; idx++) {
+		LOG_INF("adc ch: %d, for acpi sen: %d",
+			therm_sensor_tbl[idx].adc_ch,
+			therm_sensor_tbl[idx].acpi_loc);
+
+		adc_ch_bits |= BIT(therm_sensor_tbl[idx].adc_ch);
+	}
+
+	LOG_INF("adc ch sensors bit: %x", adc_ch_bits);
+	if (thermal_sensors_init(adc_ch_bits)) {
+		LOG_WRN("Thermal Sensor module init failed!");
+	} else {
+		thermal_initialized = true;
+	}
+
+}
+
+static uint8_t mapped_fan_speed(void)
+{
+	uint8_t cpu_fan_speed;
+
+	switch(cpu_temp) {
+	case 00 ... 19 :
```

Comment:
What do you think of handling cpu_temp =0 as separate case? Specially, when CPU temp is not read (PECI issue / any other reason), then leaving the fan off may heat-up / damage CPU.

Reply:
Sounds like a good idea

Reply:
Seems we have some handling when PECI has issue, we return safe temperature on failure.
Do you still see the need to add the additional case?

Reply:
Then no!